### PR TITLE
JAMES-2989 StoreMessageIdManager should comply to fetchGroup read level

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
Some mappers like memory always load everything.

Moreover testing that something is not here looks brittle to me.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by rouazana** [comment]
> Moreover testing that something is not here looks brittle to me.

I meant mostly for FetchGroupConverter.

**Comment by chibenwa** [comment]
```

[657026f0ee26100fcbec4105b8958a88c37daf76] [ERROR]   CassandraDownloadCucumberTest » ArrayIndexOutOfBounds Index 0 out of bounds fo...
[657026f0ee26100fcbec4105b8958a88c37daf76] [ERROR]   CassandraDownloadCucumberTest » ArrayIndexOutOfBounds Index 0 out of bounds fo...
[657026f0ee26100fcbec4105b8958a88c37daf76] [ERROR]   CassandraDownloadCucumberTest » ArrayIndexOutOfBounds Index 0 out of bounds fo...
[657026f0ee26100fcbec4105b8958a88c37daf76] [ERROR]   CassandraMessageSharingTest » ArrayIndexOutOfBounds Index 0 out of bounds for ...

[657026f0ee26100fcbec4105b8958a88c37daf76] java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:60)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.retrieveAndSaveAttachmentDetails(DownloadStepdefs.java:159)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.appendMessageWithAttachmentToMailbox(DownloadStepdefs.java:129)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at ✽."usera@domain.tld" mailbox "INBOX" contains a message "m1" with an attachment "a1"(cucumber/DownloadEndpoint.feature:28)
[657026f0ee26100fcbec4105b8958a88c37daf76] 
[657026f0ee26100fcbec4105b8958a88c37daf76] java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:60)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.retrieveAndSaveAttachmentDetails(DownloadStepdefs.java:159)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.appendMessageWithAttachmentToMailbox(DownloadStepdefs.java:129)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at ✽."alice@domain.tld" mailbox "INBOX" contains a message "1" with an attachment "2"(cucumber/DownloadGet.feature:31)
[657026f0ee26100fcbec4105b8958a88c37daf76] 
[657026f0ee26100fcbec4105b8958a88c37daf76] java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:60)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.retrieveAndSaveAttachmentDetails(DownloadStepdefs.java:159)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at org.apache.james.jmap.draft.methods.integration.cucumber.DownloadStepdefs.appendMessageWithAttachmentToMailbox(DownloadStepdefs.java:129)
[657026f0ee26100fcbec4105b8958a88c37daf76] 	at ✽."username@domain.tld" mailbox "INBOX" contains a message "1" with an attachment "2"(cucumber/DownloadPost.feature:35)

```

**Comment by Arsnael** [comment]
Need rebase @chibenwa

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MessageResultImpl.java [hunk 1]
```diff
@@ -232,7 +233,7 @@ public Content getBody() {
         }
 
         public void setBody(Content body) {
-            content = content | FetchGroup.BODY_CONTENT;
+            content = content | FetchGroup.BODY_CONTENT_MASK;
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/MessageResultImpl.java]: FetchGroup.FULL_CONTENT.or(content) ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java [hunk 2]
```diff
@@ -100,17 +100,17 @@ public static MessageResult loadMessageResult(MailboxMessage message, FetchGroup
             if (fetchGroup != null) {
                 int content = fetchGroup.content();
 
-                if ((content & FetchGroup.HEADERS) > 0) {
-                    content -= FetchGroup.HEADERS;
+                if ((content & FetchGroup.HEADERS_MASK) > 0) {
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java]: all those byte operation `&` and `-` should be domain method in the `FetchGroup` class

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java [hunk 3]
```diff
@@ -100,17 +100,17 @@ public static MessageResult loadMessageResult(MailboxMessage message, FetchGroup
...
+                    content -= FetchGroup.HEADERS_MASK;
                 }
-                if ((content & FetchGroup.BODY_CONTENT) > 0) {
-                    content -= FetchGroup.BODY_CONTENT;
+                if ((content & FetchGroup.BODY_CONTENT_MASK) > 0) {
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java]: zero should be a constant in FetchGroup. `NO_MASK` ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java [hunk 4]
```diff
@@ -19,42 +19,42 @@
...
      */
-    public static MessageMapper.FetchType getFetchType(MessageResult.FetchGroup group) {
+    public static MessageMapper.FetchType getFetchType(FetchGroup group) {
         int content = group.content();
         boolean headers = false;
         boolean body = false;
         boolean full = false;
 
-        if ((content & MessageResult.FetchGroup.HEADERS) > 0) {
+        if ((content & FetchGroup.HEADERS_MASK) > 0) {
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java]: idem

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java [hunk 5]
```diff
@@ -56,6 +56,14 @@
             full = true;
             content -= FetchGroup.MIME_DESCRIPTOR_MASK;
         }
+        if ((content & FetchGroup.MIME_CONTENT_MASK) > 0) {
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java]: idem

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java [hunk 6]
```diff
@@ -74,4 +59,8 @@
             return MessageMapper.FetchType.Metadata;
         }
     }
+
+    private static boolean hasMask(FetchGroup group, int mask) {
+        return (group.content() & mask) > 0;
```
Comment (remk) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/FetchGroupConverter.java]: use the `NO_MASK` constant

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/ResultUtils.java]: Let's take this refactoring another step further then :-)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MessageResultImpl.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/MessageResultImpl.java]: This is not stronlgly typed. Won't adress it here.

Let's keep some stuff for upcoming refactorings...

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java [hunk 7]
```diff
@@ -29,27 +29,23 @@
  */
 public class FetchGroupImpl implements MessageResult.FetchGroup {
 
-    public static final MessageResult.FetchGroup MINIMAL = new FetchGroupImpl(MessageResult.FetchGroup.MINIMAL);
+    public static final FetchGroupImpl MINIMAL = new FetchGroupImpl(MessageResult.FetchGroup.MINIMAL);
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java]: why would you want to promote FetchGroupImpl in place of FetchGroup?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java [hunk 8]
```diff
@@ -20,13 +20,11 @@
...
  */
-package org.apache.james.imap.processor.fetch;
+package org.apache.james.mailbox.model;
 
 import java.util.Arrays;
 
-import org.apache.james.mailbox.model.MessageResult;
-
-final class MimePathImpl implements MessageResult.MimePath {
+public class MimePathImpl implements MessageResult.MimePath {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java]: why did you remove final? Do you intend to make it extendable?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java [hunk 9]
```diff
@@ -28,7 +28,6 @@
     private final int[] positions;
 
     public MimePathImpl(int[] positions) {
-        super();
         this.positions = positions;
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java]: should we copy the array to make it immutable?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java [hunk 10]
```diff
@@ -52,17 +51,9 @@ public final int hashCode() {
...
-            } else {
-                builder.append('.');
-            }
-            builder.append(position);
-        }
-        return builder.toString();
+    @Override
+    public final String toString() {
+        return "MIMEPath:"
+            + Arrays.toString(positions);
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MimePathImpl.java]: the previous toString was much better as it's displaying a path dot-separated `a.b.c`

File: protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java [hunk 11]
```diff
@@ -188,10 +188,10 @@ protected FetchGroup getFetchGroup(FetchData fetch) {
         FetchGroup result = FetchGroup.MINIMAL;
 
         if (fetch.isEnvelope()) {
-            result.or(FetchGroup.HEADERS_MASK);
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java]: as far as I can tell, this class is not covered by tests, I thus don't agree with this refactoring.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java (hunk 7 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java]: I believe that arguments of our api should not be interfaces, especially when we expect the caller to insranciate it.

I can not recall how many time i asked myself: how to build a fetch group?

And I don t see the value of having fetchGroup and fetchgroupimpl being separate entities.

File: protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java (hunk 11 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java]: There are PLENTY of tests, MPT IMAP is doing an amazing job for fetch, every fetch level, on every implementation, and with a good message sample.

I feel comfortable! That might not be tests that you like but they are doing their job!

**Comment by Arsnael** [comment]
```
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java:[44,38] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   class FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: package org.apache.james.mailbox.model
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java:[58,38] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   class FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: package org.apache.james.mailbox.model
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageMetadataViewFactoryTest.java:[38,38] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   class FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: package org.apache.james.mailbox.model
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java:[88,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageHeaderViewFactoryTest
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java:[116,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageHeaderViewFactoryTest
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java:[115,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageFullViewFactoryTest
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java:[149,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageFullViewFactoryTest
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageMetadataViewFactoryTest.java:[81,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageMetadataViewFactoryTest
[b327f6188fe3da6c6c3964762c725af2a5a85099] [ERROR] /james-project/server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageMetadataViewFactoryTest.java:[100,69] cannot find symbol
[b327f6188fe3da6c6c3964762c725af2a5a85099]   symbol:   variable FetchGroupImpl
[b327f6188fe3da6c6c3964762c725af2a5a85099]   location: class org.apache.james.jmap.draft.model.message.view.MessageMetadataViewFactoryTest
```

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java [hunk 12]
```diff
@@ -112,7 +112,7 @@ void setUp() throws Exception {
     @Test
     void fromMessageResultsShouldReturnCorrectView() throws Exception {
         List<MessageResult> messages = messageIdManager
-            .getMessages(ImmutableList.of(message1.getMessageId()), FetchGroupImpl.MINIMAL, session);
+            .getMessages(ImmutableList.of(message1.getMessageId()), FetchGroup.MINIMAL, session);
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactoryTest.java]: Should be FULL

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java [hunk 13]
```diff
@@ -85,7 +85,7 @@ void setUp() throws Exception {
     @Test
     void fromMessageResultsShouldReturnCorrectView() throws Exception {
         List<MessageResult> messages = messageIdManager
-            .getMessages(ImmutableList.of(message1.getMessageId()), FetchGroupImpl.MINIMAL, session);
+            .getMessages(ImmutableList.of(message1.getMessageId()), FetchGroup.MINIMAL, session);
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/draft/model/message/view/MessageHeaderViewFactoryTest.java]: HEADERS

**Comment by Arsnael** [comment]
```
[ERROR] src/main/java/org/apache/james/mailbox/model/MimePath.java:[30,8] (imports) UnusedImports: Unused import - com.google.common.collect.ImmutableList.
```

File: server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java [hunk 14]
```diff
@@ -43,47 +42,10 @@
...
-        public Set<PartContentDescriptor> getPartContentDescriptors() {
-            return new HashSet<>();
-        }
-    }
-
-    private static final FetchGroup FULL_GROUP = new POP3FetchGroup() {
-
-        @Override
-        public int content() {
-            return BODY_CONTENT | HEADERS;
```
Comment (trantienduchn) [inline server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java]: can you confirm `BODY_CONTENT | HEADERS` = `FetchGroup.FULL_CONTENT`?

Comment (chibenwa) [reply server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java]: Logically speaking yes!

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java (hunk 7 above)
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java]: Either we decide there's no need for the interface at all or we need to pass the interface around, not the implementation.

Of course, in this case I don't thing there's any value in having an interface

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroupImpl.java]: For API parameters, I'm clearly in favor of having POJO.

File: protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java (hunk 11 above)
Comment (mbaechler) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java]: I don't dislike MPT tests however I believe a lot in the "test pyramid" and we should not rely on high level tests to cover what unit tests should. The first reason is: you have no way to know that the test coverage is good enough whereas with unit test it's doable.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    @VisibleForTesting
+    FetchGroup(int content, ImmutableSet<PartContentDescriptor> partContentDescriptors) {
+        this.content = content;
+        this.partContentDescriptors = partContentDescriptors;
+    }
+
+    /**
+     * Contents to be fetched. Composed bitwise.
+     *
+     * @return bitwise description
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: bitwise can be said of an operation but not of a value as far as I know.
I think it's rather a "bit field"

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+     * @see #FULL_CONTENT_MASK
+     * @see #BODY_CONTENT_MASK
+     * @see #MIME_HEADERS_MASK
+     * @see #MIME_CONTENT_MASK
+     */
+    public int content() {
+        return content;
+    }
+
+    public FetchGroup or(int content) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: what about : `with` ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    /**
+     * Adds content for the particular part.
+     * 
+     * @param path
+     *            <code>MimePath</code>, not null
+     * @param content
+     *            bitwise content constant
+     */
+    public FetchGroup addPartContent(MimePath path, int content) {
+        PartContentDescriptor newContent = partContentDescriptors.stream()
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: can you extract that to a submethod, I can't understand what it actually does, naming the search operation would probably help

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+            .filter(descriptor -> path.equals(descriptor.path()))
+            .findFirst()
+            .orElse(new PartContentDescriptor(path))
+            .or(content);
+
+        return new FetchGroup(this.content, ImmutableSet.<PartContentDescriptor>builder()
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]:
```suggestion
        return new FetchGroup(this.content, 
        Stream.concat(
        partContentDescriptors.stream()
                .filter(descriptor -> descriptor.path().equals(path)),
                Stream.of(newContent))
                .collect(Guavate.toImmutableSet());
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            .addAll(partContentDescriptors.stream()
+                .filter(descriptor -> descriptor.path().equals(path))
+                .collect(Guavate.toImmutableSet()))
+            .add(newContent)
+            .build());
+    }
+
+    public boolean hasMask(int mask) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: this method is crazy: how am I supposed to know what it does?
Do you expect to check the `content` bitfield contains at least the given mask?
Does it take any mask?
Only a power of two?
Do we check it contains at least one of the bits passed?
To be honest, I have bitfield but I can live with it. However, I don't think it's wise to expose public methods to do bitwise operations without a very good documentation + an extensive testsuite.
Given how costly it will be to implement, turning all this mess into a EnumSet would be much much faster.
I can even write it if you want to.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+        return (content & mask) > NO_MASK;
+    }
+
+    public boolean hasOnlyMasks(int... masks) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: My brain feels too small for such kind of code. Please try to write code that I'm smart-enough to read.

File: protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java (hunk 11 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/fetch/FetchProcessor.java]: Then... lets delay fetchGroup immutability, linked to FetchParser tests to another PR to not delay this changeset any longer.

Will contribute it tomorrow.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java (hunk 15 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: Pfff will correct that javadoc. Tomorrow.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java (hunk 19 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: To be really honest I want to do a full rewritte with an enumset, but I dont want to do it in this pr that is already way too big.

Can we live with bitewise descriptor in the short term, until I can find time to contribute that refactoring?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java (hunk 20 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: Again enumset should be better.

Should I get the changeset bigger?

I agree to get rid of that complicated bitewise code. The question imo is 'when'. To which I answer "after I delivered value". I can live with bitewise stuff a few more days.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java (hunk 18 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/FetchGroup.java]: No we need to update the stored partContent with or.

**Comment by mbaechler** [comment]
Please have a look at this very small refactoring that drop the bitfield usage: https://github.com/linagora/james-project/pull/2929 and let me know if we can make this change.

**Comment by chibenwa** [comment]
Looks great to me, thanks.

**Comment by chibenwa** [comment]
> Please have a look at this very small refactoring that drop the bitfield usage: #2929 and let me know if we can make this change.

I propose the following **tactical choice**: let's merge this first, then I gonna backport your refactoring. Ok?

**Comment by Arsnael** [comment]
```
[8e256234cc5cd8e2fac34f3cd990eaa9ff253afb] [INFO] There is 1 error reported by Checkstyle 8.23 with checkstyle.xml ruleset.
[8e256234cc5cd8e2fac34f3cd990eaa9ff253afb] [ERROR] src/main/java/org/apache/james/mailbox/model/MimePath.java:[30,8] (imports) UnusedImports: Unused import - com.google.common.collect.ImmutableList.
```

**Comment by Arsnael** [comment]
Actually I did that checkstyle error comment yesterday, and pushed a fix... It seems you didn't fetch it when reworking on the branch and did a force-push, discarding it. Please be careful :)

**Comment by chibenwa** [comment]
Merged.
